OrbitCaptureClient/CaptureEventProcessor.cpp

```diff
@@ -340,3 +428,254 @@ void CaptureEventProcessor::SendTracepointInfoToListenerIfNecessary(
+      command_buffer_timer.set_processor(-1);
+      command_buffer_timer.set_thread_id(thread_id);
+      command_buffer_timer.set_type(TimerInfo::kGpuCommandBuffer);
+      command_buffer_timer.set_user_data_key(command_buffer_text_key);
+      capture_listener_->OnTimer(command_buffer_timer);
+    }
+  }
+}
+
+void CaptureEventProcessor::ProcessGpuDebugMarkers(
+    const GpuQueueSubmission& gpu_queue_submission, const GpuJob& matching_gpu_job,
+    const std::optional<GpuCommandBuffer>& first_command_buffer, const std::string& timeline) {
+  if (gpu_queue_submission.completed_markers_size() == 0) {
+    return;
+  }
+  std::string timeline_marker = timeline + "_marker";
+  uint64_t timeline_marker_hash = GetStringHashAndSendToListenerIfNecessary(timeline_marker);
+
+  const auto& submission_meta_info = gpu_queue_submission.meta_info();
+  const int32_t submission_thread_id = submission_meta_info.tid();
+  uint64_t submission_pre_submission_cpu_timestamp =
+      submission_meta_info.pre_submission_cpu_timestamp();
+  uint64_t submission_post_submission_cpu_timestamp =
+      submission_meta_info.post_submission_cpu_timestamp();
+
+  for (const auto& completed_marker : gpu_queue_submission.completed_markers()) {
+    CHECK(first_command_buffer != std::nullopt);
+    TimerInfo marker_timer;
+
+    // If we've recorded the submission that contains the begin marker, we'll retrieve this
+    // submission from our mappings, and set the markers begin time accordingly.
+    // Otherwise, we will use the capture start time as begin.
+    if (completed_marker.has_begin_marker()) {
+      const auto& begin_marker_info = completed_marker.begin_marker();
+      const auto& begin_marker_meta_info = begin_marker_info.meta_info();
+      const int32_t begin_marker_thread_id = begin_marker_meta_info.tid();
+      uint64_t begin_marker_post_submission_cpu_timestamp =
+          begin_marker_meta_info.post_submission_cpu_timestamp();
+      uint64_t begin_marker_pre_submission_cpu_timestamp =
+          begin_marker_meta_info.pre_submission_cpu_timestamp();
+
+      const GpuQueueSubmission* matching_begin_submission = nullptr;
+      if (submission_pre_submission_cpu_timestamp == begin_marker_pre_submission_cpu_timestamp &&
+          submission_post_submission_cpu_timestamp == begin_marker_post_submission_cpu_timestamp &&
+          submission_thread_id == begin_marker_thread_id) {
+        matching_begin_submission = &gpu_queue_submission;
+      } else {
+        matching_begin_submission = FindMatchingGpuQueueSubmission(
+            begin_marker_thread_id, begin_marker_post_submission_cpu_timestamp);
+      }
+
+      // Note, we receive submissions of a single queue in order (by CPU submission time). So if
+      // there is no matching "begin submission", the "begin" was submitted before the "end" and
+      // we lost the record of the "begin's" submission (which should not happen).
+      CHECK(matching_begin_submission != nullptr);
```

Comment:
Are we sure this should be a `CHECK` and not handled gracefully?

Reply:
But doesn't this condition depend on a lot of assumptions down the line, in particular both on the Vulkan layer and on the communication?

Reply:
If I don't miss anything, it has the assumption that the layer sends the events in-order (for a given queue) and that the communication does neither get the order wrong nor drop messages. 
I agree that these assumptions to the communication are kind of strong, but I feels like there is actually something really going badly wrong if they don't hold anymore.
In general, I think it is a good scheme to make assumptions explicit, and fail if don't meet them.

Reply:
I have to say that still I'm a bit unsure, but I'm fine with leaving the CHECK, at least for now. Maybe Ronald has an opinion?

Reply:
@ronaldfw WDYT?

Reply:
I don't think we have as strong assertions elsewhere about the order, do we? I would somewhat be in favor of logging this as an error if we are not really sure that the condition must be true. Assertions are for cases where you know something must hold and if it doesn't you know something bad has happened. Are we sure this condition must hold? I don't have enough context to understand this fully.

If you are sure that this cannot happen (that is, it could be proven to never happen, assuming all other code is correct) or that we cannot recover from this error, then the CHECK is appropriate. Otherwise, error logging would be better.

Reply:
Alternatively, leave the CHECK while we are developing and see if we ever see this problem occur. Re-evaluate when we ship (though, easy to forget to do so).

Reply:
As far as I know, this is the only situation on the Client where we depend on the order being correct, which is why have no other places where we have assertions on the order being correct. From the reset of the code base, I am not 100% sure. 
On the Layer side, we have code to ensure the order of sending out is correct and also have test for that. 

So I'd like to ensure that this really the case, and that I not missed something.

Reply:
>Alternatively, leave the CHECK while we are developing and see if we ever see this problem occur. Re-evaluate when we ship (though, easy to forget to do so).

I created http://b/177652631 for this and made it a blocker so that we don't forget about it.
For now, I like to see how it performs.